src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,331 @@
+    ack_receiver: Receiver<u64>,
+) {
+    let mut ack_counter = 0;
+
+    let ack_ids_clone = ack_ids.clone();
+    let request_stream = async_stream::stream! {
+        yield StreamingPullRequest {
+            subscription: subscription_id.clone(),
+            ack_ids: vec![],
+            modify_deadline_seconds: vec![],
+            modify_deadline_ack_ids: vec![],
+            stream_ack_deadline_seconds: i32::try_from(ack_deadline.as_secs()).unwrap_or(10),
+            client_id: "".to_string(),
+            max_outstanding_messages: i64::try_from(QSIZE.load(Ordering::Relaxed)).unwrap_or(128),
+            max_outstanding_bytes: 0
+        };
+
+        while let Ok(pull_id) = ack_receiver.recv().await {
+            if let (Some((_, ack_id))) = ack_ids_clone.remove(&pull_id) {
+                yield StreamingPullRequest {
+                    subscription: "".to_string(),
+                    ack_ids: vec![ack_id],
+                    modify_deadline_seconds: vec![],
+                    modify_deadline_ack_ids: vec![],
+                    stream_ack_deadline_seconds: 0,
+                    client_id: "".to_string(),
+                    max_outstanding_messages: 0,
+                    max_outstanding_bytes: 0
+                };
+            } else {
+                warn!("Did not find an ACK ID for pull_id: {}", pull_id);
+            }
+        }
+    };
+    let stream = client.streaming_pull(request_stream).await;
+
+    let mut stream = match stream {
+        Ok(stream) => stream.into_inner(),
+        Err(e) => {
+            error!("Failed to send a streaming pull request: {}", e);
+            return;
+        }
+    };
+
+    while let Some(response) = stream.next().await {
+        let response = match response {
+            Ok(x) => x,
+            Err(e) => {
+                // We can't receive from the stream, exit the task, so the main task can reconnect
+                warn!("Failed to read from stream, exiting client task... {}", e);
+
+                return;
+            }
+        };
+
+        for message in response.received_messages {
+            let ReceivedMessage {
+                ack_id,
+                message: msg,
+                ..
+            } = message;
+
+            if let Some(PubsubMessage { data, .. }) = msg {
+                ack_counter += 1;
+                ack_ids.insert(ack_counter, ack_id);
+                if let Err(e) = sender.send(Ok((ack_counter, data))).await {
+                    error!("Failed to send a PubSub message to the main task: {}", e);
+
+                    // If we can't send to the main task, disconnect and let it restart
+                    return;
+                }
+            }
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Source for GSubSource {
+    async fn connect(&mut self, _ctx: &SourceContext, _attempt: &Attempt) -> Result<bool> {
+        let url = Url::<HttpsDefaults>::parse(self.config.endpoint.as_str())?;
```

Comment:
this could/should be done in the `ConnectorBuilder` above. Then we get errors for an invalid config as early as possible.